### PR TITLE
tier based ID rework

### PR DESCRIPTION
**Pull request by TheChosenEvilOne**
## About The Pull Request
This changes all accesses to have a tier from 1 to 5 which tells what is the required tier of ID card to add that access on, for example you can't give an assistant (tier 1 ID card) ID console access (tier 5 access), but you can give maintenance access (tier 1) to a scientist (tier 2 ID card). Due to there being jobs that require special access ID cards also have additional access chips (AA-chips, I only named them that because of the AA), each job that requires access with higher tier than the ID card they have will on roundstart spawn with an AA-chip that contains the access they require. AA-chips can also be removed with a screwdriver and inserted to any ID card that has chip slots by hand (this means you can steal ID console access from the HoP with a screwdriver, because the HoP's ID card is tier 4 and ID console access is tier 5).

The card tiers go as follows:
Tier 0 - Prisoner
Tier 1 - Assistant
Tier 2 - All jobs that are not mentioned
Tier 3 - QM, Warden, Virologist
Tier 4 - CE, CMO, HoP, HoS, RD
Tier 5 - Captain

## Why It's Good For The Game
The current ID trim thing is bad and limiting for no reason.

## Changelog
:cl:
add: ID card tiers, you can only add access to ID cards if the access' tier is lower or equal to the ID card's tier.
add: additional access chips, every job that has a lower tier ID card than some accesses they require spawn with an additional access chip in the ID card which contain the access required. You can use a screwdriver to remove the chip and add a chip in an ID card by clicking an ID card with it.
del: Removed ID trims.
balance: Agent ID card no longer has a limit on how many access levels it can hold.
/:cl:

### CODE REVIEW

File: _maps/map_files/Deltastation/DeltaStation2.dmm [hunk 1]
```diff
@@ -45875,7 +45875,6 @@
 /obj/effect/turf_decal/tile/neutral{
 	dir = 8
 	},
-/obj/item/paper/fluff/ids_for_dummies,
```
Comment (Canohaaaaaaa) [inline _maps/map_files/Deltastation/DeltaStation2.dmm:45878]: Do you have a replacement for this ?

File: code/game/machinery/computer/communications.dm [hunk 2]
```diff
@@ -132,7 +130,7 @@
 					to_chat(usr, span_warning("You need to swipe your ID!"))
 					playsound(src, 'sound/machines/terminal_prompt_deny.ogg', 50, FALSE)
 					return
-				if (!(ACCESS_CAPTAIN in id_card.access))
+				if (!(ACCESS_CAPTAIN in id_card.GetAccess()))
```
Comment (Canohaaaaaaa) [inline code/game/machinery/computer/communications.dm:133]: `snake_case` the ID procs please

File: code/game/machinery/computer/communications.dm [hunk 3]
```diff
@@ -313,7 +311,8 @@
 
 			if (obj_flags & EMAGGED)
 				authenticated = TRUE
-				authorize_access = SSid_access.get_region_access_list(list(REGION_ALL_STATION))
+				// TODO:
+				// authorize_access = SSid_access.get_region_access_list(list(REGION_ALL_STATION))
```
Comment (Canohaaaaaaa) [inline code/game/machinery/computer/communications.dm:315]: TODO leftover

File: code/game/machinery/porta_turret/portable_turret.dm [hunk 4]
```diff
@@ -547,7 +547,7 @@ DEFINE_BITFIELD(turret_flags, list(
 			return 10
 
 	if(turret_flags & TURRET_FLAG_AUTH_WEAPONS) //check for weapon authorization
-		if(isnull(perp.wear_id) || istype(perp.wear_id.GetID(), /obj/item/card/id/advanced/chameleon))
+		if(isnull(perp.wear_id) || istype(perp.wear_id.GetID(), /obj/item/card/id/chameleon))
```
Comment (Canohaaaaaaa) [inline code/game/machinery/porta_turret/portable_turret.dm:550]: `snake_case`

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 5]
```diff
@@ -534,7 +534,7 @@
 				open()
 			else
 				req_access = list()
-				req_access += pick(SSid_access.get_region_access_list(list(REGION_ALL_STATION)))
+				//TODO: req_access += pick(SSid_access.get_region_access_list(list(REGION_ALL_STATION)))
```
Comment (Canohaaaaaaa) [inline code/game/objects/structures/crates_lockers/closets.dm:537]: TODO leftover

File: code/controllers/subsystem/id_access.dm [hunk 6]
```diff
@@ -335,143 +171,49 @@ SUBSYSTEM_DEF(id_access)
...
 	var/list/id_card_access = id_card.access
 	for(var/access in id_card_access)
-		if(flags_by_access["[access]"] >= access_flag)
+		if (tiers_by_access["[access]"] >= tier)
 			tally++
 
 	return tally
+
+/datum/controller/subsystem/id_access/proc/apply_card_access(obj/item/card/id/id, card_access, chip = FALSE, force = FALSE)
+	var/datum/card_access/C = card_access_instances[card_access]
```
Comment (Canohaaaaaaa) [inline code/controllers/subsystem/id_access.dm:206]: Descriptive var name

File: code/game/machinery/PDApainter.dm [hunk 7]
```diff
@@ -40,334 +30,113 @@
...
 	if(machine_stat & BROKEN)
 		if(O.tool_behaviour == TOOL_WELDER && !user.combat_mode)
 			if(!O.tool_start_check(user, amount=0))
 				return
-			user.visible_message(span_notice("[user] is repairing [src]."), \
-							span_notice("You begin repairing [src]..."), \
-							span_hear("You hear welding."))
+			user.visible_message("<span class='notice'>[user] is repairing [src].</span>", \
+							"<span class='notice'>You begin repairing [src]...</span>", \
+							"<span class='hear'>You hear welding.</span>")
```
Comment (Canohaaaaaaa) [inline code/game/machinery/PDApainter.dm:68]:
```suggestion
			user.visible_message(span_notice("[user] is repairing [src]."), \
							span_notice("You begin repairing [src]..."), \
							span_hear("You hear welding."))
```

File: code/game/machinery/PDApainter.dm [hunk 8]
```diff
@@ -40,334 +30,113 @@
...
 			if(O.use_tool(src, user, 40, volume=50))
 				if(!(machine_stat & BROKEN))
 					return
-				to_chat(user, span_notice("You repair [src]."))
+				to_chat(user, "<span class='notice'>You repair [src].</span>")
```
Comment (Canohaaaaaaa) [inline code/game/machinery/PDApainter.dm:72]:
```suggestion
				to_chat(user, span_notice("You repair [src]."))
```

File: code/game/machinery/PDApainter.dm [hunk 9]
```diff
@@ -40,334 +30,113 @@
...
+	else if(default_unfasten_wrench(user, O))
+		power_change()
 		return
 
-	if(istype(O, /obj/item/card/id))
-		if(stored_id_card)
-			to_chat(user, span_warning("There is already an ID card inside!"))
+	else if(istype(O, /obj/item/pda))
+		if(storedpda)
+			to_chat(user, "<span class='warning'>There is already a PDA inside!</span>")
```
Comment (Canohaaaaaaa) [inline code/game/machinery/PDApainter.dm:86]:
```suggestion
			to_chat(user, span_warning("There is already a PDA inside!"))
```

File: code/game/machinery/PDApainter.dm [hunk 10]
```diff
@@ -40,334 +30,113 @@
...
-
-			for(var/path in pda_types)
-				if(pda_types[path] == selection)
-					pda_path = path
-					break
-
+	if(storedpda)
+		if(machine_stat & BROKEN) //otherwise the PDA is stuck until repaired
+			ejectpda()
+			to_chat(user, "<span class='info'>You manage to eject the loaded PDA.</span>")
```
Comment (Canohaaaaaaa) [inline code/game/machinery/PDApainter.dm:108]:
```suggestion
			to_chat(user, span_info("You manage to eject the loaded PDA."))
```

File: code/game/machinery/PDApainter.dm [hunk 11]
```diff
@@ -40,334 +30,113 @@
...
-
-			stored_id_card.clear_account()
-
-			return TRUE
+				storedpda.icon = initial(pda_path.icon)
+			storedpda.icon_state = initial(pda_path.icon_state)
+			storedpda.desc = initial(pda_path.desc)
+			ejectpda()
+	else
+		to_chat(user, "<span class='warning'>[src] is empty!</span>")
```
Comment (Canohaaaaaaa) [inline code/game/machinery/PDApainter.dm:126]:
```suggestion
		to_chat(user, span_warning("[src] is empty!"))
```

File: code/game/machinery/PDApainter.dm [hunk 12]
```diff
@@ -40,334 +30,113 @@
...
-/// Engineering departmental variant. Limited to PDAs defined in the SSid_access.sub_department_managers_tgui data structure.
-/obj/machinery/pdapainter/engineering
-	name = "\improper Engineering PDA & ID Painter"
-	target_dept = REGION_ENGINEERING
+	if(storedpda)
+		storedpda.forceMove(drop_location())
+		storedpda = null
+		update_appearance()
+	else
+		to_chat(usr, "<span class='warning'>[src] is empty!</span>")
```
Comment (Canohaaaaaaa) [inline code/game/machinery/PDApainter.dm:142]:
```suggestion
		to_chat(usr, span_warning("[src] is empty!"))
```

**Review by Canohaaaaaaa — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/id_access.dm (hunk 6 above)
Comment (TheChosenEvilOne) [reply code/controllers/subsystem/id_access.dm:206]: watch this

File: code/controllers/subsystem/id_access.dm [hunk 13]
```diff
@@ -335,143 +171,49 @@ SUBSYSTEM_DEF(id_access)
...
+		if (!force && id.access_tier < text2num(tier))
+			if (chip)
+				chip_access.Add(C.access[tier])
+			continue
+		id.access.Add(C.access[tier])
+
+	if (chip_access.len)
+		var/obj/item/card_access_chip/roundstart/AA = new(id, C.assignment)
+		AA.access = chip_access
+		id.apply_access_chip(AA)
```
Comment (TheChosenEvilOne) [inline code/controllers/subsystem/id_access.dm:219]:
```suggestion
/datum/controller/subsystem/id_access/proc/apply_card_access(obj/item/card/id/id, card_access_path, chip = FALSE, force = FALSE)
	var/datum/card_access/card_access = card_access_instances[card_access_path]
	var/list/chip_access = list()
	id.assignment = card_access.assignment
	for (var/tier in card_access.access)
		if (!force && id.access_tier < text2num(tier))
			if (chip)
				chip_access.Add(card_access.access[tier])
			continue
		id.access.Add(card_access.access[tier])

	if (chip_access.len)
		var/obj/item/card_access_chip/roundstart/AA = new(id, card_access.assignment)
		AA.access = chip_access
		id.apply_access_chip(AA)
```

**Comment by kriskog** [comment]
Some stuff:
Requires a chip for weapons permit. This is something the HOS for instance should be able to grant anyone.
No chip slot icon on IDs until you put one on. Removing the chip leaves the slot icon.
No extra chips on station? Was alone one round and had to grab spare, but spare has no chip. Forced to abandon own ID (including bank account) for access. What if the HOP or Captain wishes to promote someone to a head role? Replace someone after an unrecoverable suicide? Prisoners and assistants should be able to get a job and that jobs access if thats how it plays out. You need to figure something out for this. A limited amount of Tier 4 capable chips somewhere secure at least.
Didn't check, but have heads retained the ability to give access inside their departments?

**Comment by TheChosenEvilOne** [comment]
well, I already wrote a response to this but forgot to post it so, here I go again.
> * Requires a chip for weapons permit. This is something the HOS for instance should be able to grant anyone.

Giving HoS a way to grant people weapons permit would be good, but I am not sure how it should work. I could give the HoS a box of weapons permit chips or a way to produce weapons permit chips.

> * No chip slot icon on IDs until you put one on. Removing the chip leaves the slot icon.

Fixed this, IDs didn't update overlays when created.

> * No extra chips on station? Was alone one round and had to grab spare, but spare has no chip. Forced to abandon own ID (including bank account) for access. What if the HOP or Captain wishes to promote someone to a head role? Replace someone after an unrecoverable suicide? Prisoners and assistants should be able to get a job and that jobs access if thats how it plays out. You need to figure something out for this. A limited amount of Tier 4 capable chips somewhere secure at least, and Tier 3 chips should be abundant.

I need to figure out how getting higher tier IDs and chips should work, right now what I have been thinking about is that tier 1 and 2 IDs and tier 3 chips should be printable. I could add a box of tier 4 chips in HoP (captain?) locker and possibly allow buying them. Tier 3 IDs as well could be something you could get from cargo.

> * Didn't check, but have heads retained the ability to give access inside their departments?

I added this back, I removed it without realising.

**Comment by TheChosenEvilOne** [comment]
I'd say this is about done, I don't really have anything to add to this.

**Pull request merged**